carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java

```diff
@@ -367,37 +373,38 @@ public MapResult map(HttpRequest request, String userId, String sessionId, Backe
                 }
                 somethingMatched = somethingMatched | !selectedBackends.isEmpty();
+
                 LOG.log(Level.FINEST, "selected {0} backends for {1}, director is {2}", new Object[]{selectedBackends, request.uri(), director});
                 for (String backendId : selectedBackends) {
+                    Action selectedAction;
                     switch (action.getType()) {
-                        case ActionConfiguration.TYPE_PROXY: {
-                            BackendConfiguration backend = this.backends.get(backendId);
-                            if (backend != null && backendHealthManager.isAvailable(backendId)) {
-                                return new MapResult(
-                                        backend.getHost(),
-                                        backend.getPort(),
-                                        MapResult.Action.PROXY,
-                                        route.getId(),
-                                        action.getCustomHeaders()
-                                );
-                            }
+                        case ActionConfiguration.TYPE_PROXY:
+                            selectedAction = MapResult.Action.PROXY;
                             break;
-                        }
                         case ActionConfiguration.TYPE_CACHE:
-                            BackendConfiguration backend = this.backends.get(backendId);
-                            if (backend != null && backendHealthManager.isAvailable(backendId)) {
-                                return new MapResult(
-                                        backend.getHost(),
-                                        backend.getPort(),
-                                        MapResult.Action.CACHE,
-                                        route.getId(),
-                                        action.getCustomHeaders()
-                                );
-                            }
+                            selectedAction = MapResult.Action.CACHE;
                             break;
                         default:
                             return MapResult.NOT_FOUND(route.getId());
                     }
+
+                    BackendConfiguration backend = this.backends.get(backendId);
+                    if (backend != null && backendHealthManager.isAvailable(backendId)) {
+                        if (this.debuggingHeaderEnabled) {
+                            String routingPath = "Route-id: " + route.getId() + "; ";
+                            routingPath += "Action-id: " + action.getId() + "; ";
+                            routingPath += "Director-id: " + action.getDirector() + "; ";
+                            routingPath += "Backend-id: " + backendId;
+                            action.addCustomHeader(new CustomHeader(DEBUGGING_HEADER_ROUTING_PATH, routingPath, HeaderMode.HEADER_MODE_ADD));
```

Comment:
this is not good, "Actions" are immutable objects, you cannot add a "custom header"

Reply:
Do you think it's ok cloning the object ('cause i think it's good adding all additional headers at this higher level as CustomHeader)? Or is it better to add it in the request handler (lower level)?

Reply:
Clone = allocate...
If we can save allocations on the hot path it is better.
So I prefer not to clone